ruby/ql/lib/codeql/ruby/frameworks/ActionDispatch.qll

```diff
@@ -0,0 +1,889 @@
+     */
+    TResourceRoute(RouteBlock b, MethodCall m, string action) {
+      b.getAStmt() = m and
+      m.getMethodName() = "resource" and
+      action in ["show", "new", "edit", "create", "update", "destroy"] and
+      applyActionFilters(m, action)
+    } or
+    /**
+     * See `MatchRoute`
+     */
+    TMatchRoute(RouteBlock b, MethodCall m) { b.getAStmt() = m and m.getMethodName() = "match" }
+
+  /**
+   * A route configuration. This defines a combination of HTTP method and URL
+   * path which should be routed to a particular controller-action pair.
+   * This can arise from an explicit call to a routing method, for example:
+   * ```rb
+   * get "/photos", to: "photos#index"
+   * ```
+   * or via a convenience method like `resources`, which defines mutiple routes at once:
+   * ```rb
+   * resources :photos
+   * ```
+   */
+  abstract class Route extends TRoute {
+    /**
+     * Gets the name of a primary CodeQL class to which this route belongs.
+     */
+    string getAPrimaryQlClass() { result = "Route" }
+
+    MethodCall method;
+
+    /** Gets a string representation of this route. */
+    string toString() { result = method.toString() }
+
+    /**
+     * Gets the location of the method call that defines this route.
+     */
+    Location getLocation() { result = method.getLocation() }
+
+    /**
+     * Gets the method call that defines this route.
+     */
+    MethodCall getDefiningMethodCall() { result = method }
+
+    /**
+     * Get the last component of the path. For example, in
+     * ```rb
+     * get "/photos", to: "photos#index"
+     * ```
+     * this is `/photos`.
+     * If the string has any interpolations, this predicate will have no result.
+     */
+    abstract string getLastPathComponent();
+
+    /**
+     * Gets the HTTP method of this route.
+     * The result is one of [get, post, put, patch, delete].
+     */
+    abstract string getHTTPMethod();
+
+    /**
+     * Gets the last controller component.
+     * This is the controller specified in the route itself.
+     */
+    abstract string getLastControllerComponent();
+
+    /**
+     * Gets a component of the controller.
+     * This behaves identically to `getPathComponent`, but for controller information.
+     */
+    string getControllerComponent(int n) {
+      if n = 0
+      then result = this.getLastControllerComponent()
+      else result = this.getParentBlock().getParent(n - 1).getControllerComponent()
+    }
+
+    /**
+     * Gets the full controller targeted by this route.
+     */
+    string getController() {
+      result =
+        concat(int n |
+          this.getControllerComponent(n) != ""
+        |
+          this.getControllerComponent(n), "/" order by n desc
+        )
+    }
+
+    /**
+     * Gets the action targeted by this route.
+     */
+    abstract string getAction();
+
+    /**
+     * Gets the parent `RouteBlock` of this route.
+     */
+    abstract RouteBlock getParentBlock();
```

Comment:
`RouteBlock` is private, was that a mistake?

Reply:
I don't understand - what's the problem you're seeing here?

Reply:
What I meant is that this class is public, and it exposes a predicate that returns a value of a private type.

Reply:
Got it. I'll see if we can push this method down into `RouteImpl` and thereby keep it private.